CRM/Certificate/Token/Certificate.php

```diff
@@ -72,63 +71,57 @@ public function prefetch(TokenValueEvent $e) {
-
     $contactIds = $e->getTokenProcessor()->getContextValues('contactId');
+    $membershipIds = $e->getTokenProcessor()->getContextValues('entityId');
     $contactId = (is_array($contactIds) && !empty($contactIds[0])) ? $contactIds[0] : 0;
+    $membershipId = (is_array($membershipIds) && !empty($membershipIds[0])) ? $membershipIds[0] : 0;
 
-    // get all memberships for given contact to get min start_date and max end_date of all memberships
-    $memberships = Membership::get(FALSE)
-      ->addSelect('start_date', 'end_date')
-      ->addWhere('contact_id', '=', $contactId)
-      ->addWhere('status_id', '=', self::MEMBERSHIP_STATUS_CURRENT)
-      ->execute()
-      ->getArrayCopy();
-
-    foreach ($memberships as $membership) {
-      $startDate = $startDate === NULL || strtotime($membership['start_date']) < strtotime($startDate) ?
-        $membership['start_date'] : $startDate;
-      $endDate = $endDate === NULL || strtotime($membership['end_date']) > strtotime($endDate) ?
-        $membership['end_date'] : $endDate;
-    }
+    $membership = civicrm_api3('Membership', 'getsingle', [
+      'id' => $membershipId,
+      'contact_id' => $contactId,
+    ]);
 
-    return ['start_date' => $startDate, 'end_date' => $endDate];
+    return ['startDate' => $membership['start_date'] ?? NULL, 'endDate' => $membership['end_date'] ?? NULL];
   }
 
   /**
    * Resolve the value of ceritificate configuration token fields.
    *
    * @param CRM_Certificate_DAO_CompuCertificate $certificate
-   * @param array $membershipDates
    * @param array &$resolvedTokens
+   * @param Civi\Token\Event\TokenValueEvent $e
    */
-  private function resolveFields($certificate, array $membershipDates, &$resolvedTokens) {
-    $membershipStartTimestamp = !empty($membershipDates['start_date']) ? strtotime($membershipDates['start_date']) : '';
-    $membershipEndTimestamp = !empty($membershipDates['end_date']) ? strtotime($membershipDates['end_date']) : '';
-    $certificateValidityStartTimestamp = !empty($certificate->min_valid_from_date) ? strtotime($certificate->min_valid_from_date) : '';
-    $certificateValidityEndTimestamp = !empty($certificate->max_valid_through_date) ? strtotime($certificate->max_valid_through_date) : '';
+  private function resolveFields($certificate, &$resolvedTokens, $e) {
     $resolvedTokens['rolling_start_or_renewal_date'] = '';
-
-    if ($membershipStartTimestamp && $membershipEndTimestamp) {
-      $renewalTimestamp = strtotime($membershipDates['end_date'] . " -1 year 1 day");
-      $resolvedTokens['rolling_start_or_renewal_date'] = $renewalTimestamp > $membershipStartTimestamp
-        ? CRM_Utils_Date::customFormat(date('Y-m-d', $renewalTimestamp), '%e/%b/%Y')
-        : CRM_Utils_Date::customFormat($membershipDates['start_date'], '%e/%b/%Y');
+    $resolvedTokens['valid_from'] = '';
+    $resolvedTokens['valid_to'] = '';
+
+    if ((int) $certificate->entity === CRM_Certificate_Enum_CertificateType::MEMBERSHIPS) {
```

Comment:
I would suggest logic for Membership based tokens should here:
https://github.com/compucorp/uk.co.compucorp.certificate/blob/7d66548ebf5e19b6c5f0c477eac90de569063f8f/CRM/Certificate/Token/Membership.php

Reply:
Also, see how entity tokens are namespaced here

https://github.com/compucorp/uk.co.compucorp.certificate/blob/7d66548ebf5e19b6c5f0c477eac90de569063f8f/certificate.php#L143-L150

Reply:
as agreed on slack these are actually certificate properties so they make more sense to be in certificate class